### PR TITLE
Domains: Add advanced transfer options card

### PR DESCRIPTION
**Pull request by leonardost**
### Changes proposed in this Pull Request

**Note**: This depends on D70720-code.

This PR creates the advanced transfer options card in the new redesigned transfer page. This card contains a transfer lock toggle for the domain and a button to send the domain's authorization code to its owner - these actions were previously contained in the "Transfer to another registrar" option, which we won't have anymore.

A new action and reduce were created because the existing `DOMAIN_TRANSFER_CODE_REQUEST` one encompassed unlocking the domain and also disabling its privacy and sending the auth code via email, which will remain as separate actions now.

If the domain is currently in a transfer lock - either because it's a new domain (less than 60 days old) or because its contact info was updated and the owner chose not to opt-put of the lock - the toggle will be disabled and an explanation and a link to the relevant support docs will be shown to the user.

This is part of the domain management redesign project described in pcYYhz-m2-p2.

#### Screenshots

Old "Transfer to another registrar" page that will be removed

Screen Shot 2021-11-26 at 15 11 35

Advanced options card in desktop view

Markup on 2021-12-08 at 16:04:22

And in mobile view

Markup on 2021-12-08 at 16:05:11

When the domain is under a transfer lock, the following message will be displayed:

Screen Shot 2021-12-08 at 16 03 34

### Testing instructions

- Apply D70720-code to your sandbox if it hasn't been merged yet and sandbox `public-api`
- Build this branch locally or open the live Calypso link
- Go to "Upgrades > Domains"
- Select one of your registered domains or domain connections
- If you're in the live Calypso link, please append ?flags=domains/transfers-redesign to your URL to enable the appropriate feature flag
- Select the "Transfer your domain" options
- Please check that:
  - Advanced options card is present
  - Lock/unlock toggle works as expected
  - "Get authorization code" button works
  - Mobile view looks correct as shown in the screenshot above
  - If you selected a domain connection, ensure the advanced options do not appear, as they are only pertinent to registered domains
  - Also please check a domain which is under transfer lock

### CODE REVIEW

File: client/my-sites/domains/domain-management/transfer/transfer-page/index.tsx [hunk 1]
```diff
@@ -89,22 +99,87 @@ const TransferPage = ( props: TransferPageProps ): JSX.Element => {
...
+		);
+	};
+
+	const requestTransferCode = () => {
+		requestDomainTransferCodeOnly( selectedDomainName );
+	};
+
+	const renderAdvancedTransferOptions = () => {
+		// translators: domain transfer lock
+		const disabledLockLabel = __( 'Transfer lock off' );
```
Comment (a8ci18n) [inline client/my-sites/domains/domain-management/transfer/transfer-page/index.tsx:200]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **20** times:
`translate( 'Transfer lock: off' )` *ES Score: 12*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/advanced-transfer-options-component)

File: client/my-sites/domains/domain-management/transfer/transfer-page/index.tsx [hunk 2]
```diff
@@ -89,22 +99,87 @@ const TransferPage = ( props: TransferPageProps ): JSX.Element => {
...
+		// translators: domain transfer lock
+		const enabledLockLabel = __( 'Transfer lock on' );
```
Comment (a8ci18n) [inline client/my-sites/domains/domain-management/transfer/transfer-page/index.tsx:202]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **23** times:
`translate( 'Transfer lock: on' )` *ES Score: 11*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/advanced-transfer-options-component)

File: client/my-sites/domains/domain-management/transfer/transfer-page/index.tsx [hunk 3]
```diff
@@ -89,22 +99,87 @@ const TransferPage = ( props: TransferPageProps ): JSX.Element => {
...
+					onChange={ () => null }
+					label={ toggleLabel }
+				/>
+				<p>
+					{ __(
+						'We recommend leaving the transfer lock on, unless you want to transfer your domain to another provider.'
+					) }
+				</p>
+				<Button primary={ false } busy={ isRequestingTransferCode } onClick={ requestTransferCode }>
+					{ __( 'Get authorization code' ) }
```
Comment (a8ci18n) [inline client/my-sites/domains/domain-management/transfer/transfer-page/index.tsx:275]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **14** times:
`translate( 'Get domain authorization code' )` *ES Score: 7*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/advanced-transfer-options-component)

**Review by hambai — CHANGES_REQUESTED** [review]
I believe we can simplify this PR and not keep all those new actions, action creators and reducers in the global state. We should be able to just use local component state for `isRequestingTransferCode` and `isLockingOrUnlockingDomain`. I also think that we can safely assume that if we got a success from the api endpoint we can just update the global state for the domain lock status and not make additional api call to get the updated status?

**Comment by leonardost** [comment]
Thanks for the review @hambai! You're definitely right, not everything needs to be in redux or in the global state. I started from the existing code and didn't even think about doing things more simply, so thanks for the discussion yesterday 🙇 

Now there's only one action to update the lock status of the domain when the API call is successful, so there's a lot less boilerplate 🙂 would love if you could take another look at this.

**Review by rafaelgallani — COMMENTED** [review]
Nice job! I tested and all the visual changes were reflected, and the authorization code button also worked as expected. LGTM.

But there was a single point where I have noticed something weird:

image

I wasn't able to turn off the transfer lock for my domain. 
I believe that's _probably_ because I marked the 60-day transfer lock when I previously updated (a few days ago) the domain's contact info. There's also the transfer's card message confirming it:

image

I think it might be a good idea to disable the button when the domain is in this state, otherwise, it might get some users confused.

**Comment by leonardost** [comment]
Thanks for the review @rafaelgalani and nice catch! I'll add a check for the 60-day transfer lock 👍

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7030084

Thank you @leonardost for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7030084).

**Pull request merged**